module/zfs/arc.c

```diff
@@ -6032,6 +6032,23 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				    ARC_FLAG_PREDICTIVE_PREFETCH);
 			}
 
+			/*
+			 * If there are multiple threads reading the same block
+			 * and that block is not yet in the ARC, then only one
+			 * thread will do the physical I/O and all other
+			 * threads will wait until that I/O completes.
+			 * Synchronous reads use the b_cv whereas nowait reads
+			 * register a callback. Both are signalled/called in
+			 * arc_read_done.
+			 *
+			 * Errors of the physical I/O need to be propagated to
```

Comment:
NIT: `Errors of the logical I/O may need...` The `zio_flags` may indicate that errors should not be propagated.